File: benchmarks/primes.fs
\ Sieve of Eratosthenes benchmark
100001 constant N
create sieve N allot

: init-sieve
  sieve N 1 fill
  0 sieve c!
  0 sieve 1+ c! ;

: mark-composite ( start step -- )
  N 1- swap do
    0 sieve i + c!
  dup +loop drop ;

: sieve-pass ( i -- )
  dup dup * N < if
    sieve over + c@ if
      dup dup * swap mark-composite
    else drop then
  else drop then ;

: run-sieve
  init-sieve
  N 2 do i sieve-pass loop ;

: count-primes ( -- n )
  0 N 2 do sieve i + c@ + loop ;

: main
  run-sieve
  count-primes . cr ;

File: benchmarks/primes.c
// Sieve of Eratosthenes benchmark
#include <stdio.h>
#include <string.h>

#define N 100000

char sieve[N + 1];

int main() {
    memset(sieve, 1, sizeof(sieve));
    sieve[0] = sieve[1] = 0;

    for (int i = 2; i * i <= N; i++) {
        if (sieve[i]) {
            for (int j = i * i; j <= N; j += i) {
                sieve[j] = 0;
            }
        }
    }

    int count = 0;
    for (int i = 2; i <= N; i++) {
        if (sieve[i]) count++;
    }

    printf("%d\n", count);
    return 0;
}
